### PR TITLE
Fixed memory issues in Plot class

### PR DESCRIPTION
**Pull request by sebastianbaginski**
Fixes https://github.com/cogneco/ooc-kean/issues/819

It was even better than simple memory leaks. It was impossible to call `free` on `SvgWriter2D` if it had more than one plot data.
To be more specific, it was impossible to call `free` on two different instances of `SvgPlot` class. It crashed because of double delete/free. Reason for this was the `colorList` variable, filled here:
```ooc
	fillColorList: func {
		this colorList = VectorList<String> new()
		this colorList add("crimson")
		this colorList add("black")
		this colorList add("darkmagenta")
		this colorList add("seagreen")
		this colorList add("dodgerblue")
		this colorList add("darkorange")
		this colorList add("hotpink")
		this colorList add("lightgreen")
		this colorList add("yellowgreen")
		this colorList add("red")
	}
```
Each of the color entries was compiled into executable as string literal, so in the end every `colorList` in each `SvgPlot` pointed to the same data:
```C
        __SvgPlot_strLit20 = (void*) lang_String__makeStringLiteral("yellowgreen", 11);
        __SvgPlot_strLit19 = (void*) lang_String__makeStringLiteral("lightgreen", 10);
        __SvgPlot_strLit18 = (void*) lang_String__makeStringLiteral("hotpink", 7);
        __SvgPlot_strLit17 = (void*) lang_String__makeStringLiteral("darkorange", 10);
        __SvgPlot_strLit16 = (void*) lang_String__makeStringLiteral("dodgerblue", 10);
        __SvgPlot_strLit15 = (void*) lang_String__makeStringLiteral("seagreen", 8);
        __SvgPlot_strLit14 = (void*) lang_String__makeStringLiteral("darkmagenta", 11);
        __SvgPlot_strLit13 = (void*) lang_String__makeStringLiteral("black", 5);
        __SvgPlot_strLit12 = (void*) lang_String__makeStringLiteral("crimson", 7);

...
void SvgPlot__SvgPlot_fillColorList(SvgPlot__SvgPlot* this) {
    SvgPlot__SvgPlot___setcolorList__(this, VectorList__VectorList_new_default((lang_types__Class*)lang_String__String_class()));
    VectorList__VectorList_add(SvgPlot__SvgPlot___getcolorList__(this), (uint8_t*) &(__SvgPlot_strLit12));
    VectorList__VectorList_add(SvgPlot__SvgPlot___getcolorList__(this), (uint8_t*) &(__SvgPlot_strLit13));
...
```
This PR fixes these memory issues and some memory leaks.

### CODE REVIEW

File: source/plot/SvgPlot.ooc [hunk 1]
```diff
@@ -34,7 +34,10 @@ SvgPlot: class {
 		this datasets free()
 		this xAxis free()
 		this yAxis free()
-		this colorList free()
+		if (this colorList) {
+			this colorList _count = 0
```
Comment (sebastianbaginski) [inline source/plot/SvgPlot.ooc]: `colorList` entries are shared (as in the PR description), this is a way around that which allows to free the `colorList` anyway.

File: source/plot/SvgPlot.ooc [hunk 2]
```diff
@@ -96,7 +99,7 @@ SvgPlot: class {
 					datasets[j] color = datasets[j] colorBgra svgRGBToString()
 					datasets[j] opacity = ((datasets[j] colorBgra svgRGBAlpha()) as Float) / 255.0f
 				} else {
-					datasets[j] color = colorList[this colorCount % this colorList count]
+					datasets[j] color = colorList[this colorCount % this colorList count] clone()
```
Comment (sebastianbaginski) [inline source/plot/SvgPlot.ooc]: Each `dataset` will free its `color` (and colorList entries are shared), so need to copy it.

File: source/plot/SvgWriter2D.ooc [hunk 3]
```diff
@@ -43,13 +41,12 @@ SvgWriter2D: class {
 	init: func ~withPositioningFilename (filename: String, =numberOfPlotsHorizontally, args: ...) {
 		this init(File new(filename), args)
 	}
-	init: func ~svgPlots (file: File, svgPlots: VectorList<SvgPlot>) {
-		this init(file)
-		this svgPlots = svgPlots
```
Comment (sebastianbaginski) [inline source/plot/SvgWriter2D.ooc]: This assigned new object to `this svgPlots` without deleting old one (same below).

File: test/plot/SvgPlotTest.ooc [hunk 4]
```diff
@@ -60,8 +60,8 @@ unitCirclePlot yAxis min = -1.5
 unitCirclePlot yAxis max = 1.5
 
 // Symmetric plot
-symmetricUnitCircleData := LinePlotData2D new(unitCircle, "Unit circle")
-symmetricUnitCirclePlot := SvgPlot new(unitCircleData, "Unit circle plot with symmetic set to true")
+symmetricUnitCircleData := LinePlotData2D new(unitCircle copy(), "Unit circle")
```
Comment (sebastianbaginski) [inline test/plot/SvgPlotTest.ooc]: `LinePlotData2D` owns the vector.

File: source/plot/SvgPlot.ooc (hunk 1 above)
Comment (marcusnaslund) [inline source/plot/SvgPlot.ooc]: This seems.. weird. Why not set `freeContent` to `false` when creating the `colorList`?

Comment (sebastianbaginski) [inline source/plot/SvgPlot.ooc]: Oh yeah, I forgot that it exists

Comment (sebastianbaginski) [inline source/plot/SvgPlot.ooc]: Updated, thanks.

**Comment by sebastianbaginski** [comment]
@davidpiuva do you have time to review ? it's only few lines of minor changes

**Comment by davidpiuva** [comment]
ok

File: source/plot/SvgPlot.ooc [hunk 5]
```diff
@@ -104,7 +105,7 @@ SvgPlot: class {
 		}
 	}
 	fillColorList: func {
-		this colorList = VectorList<String> new()
+		this colorList = VectorList<String> new(10, false)
```
Comment (davidpiuva) [inline source/plot/SvgPlot.ooc]: Where are the colors freed?

Comment (sebastianbaginski) [inline source/plot/SvgPlot.ooc]: I think `false` means that they are ntot freed. They are string literals as described in the issue description. It is another weird thing in this language. I tried to reproduce this in a simple test case and 
```ooc
test := "string"
test2 := "string"
```
actually allocates two separate objects, while here they all points to static literals. I don't know what triggers each mode.

**Comment by davidpiuva** [comment]
It looks like a step in the right direction so I won't complain about the many things that are still untouched.

File: source/plot/SvgPlot.ooc (hunk 5 above)
Comment (davidpiuva) [inline source/plot/SvgPlot.ooc]: Then we need a better color representation but that can be a separate pull request since the string mess already existed.

File: source/plot/SvgPlot.ooc [hunk 6]
```diff
@@ -104,7 +105,7 @@ SvgPlot: class {
...
 		this colorList add("crimson")
```
Comment (davidpiuva) [inline source/plot/SvgPlot.ooc]: To me, RGB values say more than names that don't include the pure color's name.

Comment (sebastianbaginski) [inline source/plot/SvgPlot.ooc]: I agree, this needs cleanup.

Comment (marcusnaslund) [inline source/plot/SvgPlot.ooc]: Not here, though.

**Comment by davidpiuva** [comment]
I am done with syntax inspection.

**Comment by marcusnaslund** [comment]
Just want to say I love that you annotated your own PR, that made reviewing it so much easier.

**Comment by sebastianbaginski** [comment]
No problem, I wanted to predict possible questions :)

**Comment by sebastianbaginski** [comment]
Rebased
